lib/sigstore/models.rb

```diff
@@ -64,263 +69,182 @@ def self.from_media_type(media_type)
+        raise Error::InvalidBundle, "expected either message_signature or dsse_envelope"
       end
-
-      entry
     end
 
-    def self.from_bundle(input:, bundle:, offline:)
-      media_type = BundleType.from_media_type(bundle.media_type)
+    private
 
-      case media_type
-      when BundleType::BUNDLE_0_3
-        leaf_cert = Internal::X509::Certificate.read(bundle.verification_material.certificate.raw_bytes)
-      when BundleType::BUNDLE_0_1, BundleType::BUNDLE_0_2
-        certs = bundle.verification_material.x509_certificate_chain.certificates.map do |cert|
-          Internal::X509::Certificate.read(cert.raw_bytes)
-        end
-        raise Error::InvalidBundle, "Expected certificate chain" if certs.empty?
-
-        leaf_cert = certs.shift
-        raise Error::InvalidBundle, "Expected leaf certificate" unless leaf_cert.leaf?
-
-        certs.each do |cert|
-          raise Error::InvalidBundle, "Root CA in chain" if cert.ca?
-        end
-      else
-        raise Error::InvalidBundle, "Unsupported bundle format: #{media_type}"
-      end
-
-      case bundle.content
-      when :message_signature
-        signature = bundle.message_signature.signature
-      when :dsse_envelope
-        dsse_envelope = bundle.dsse_envelope
-      else
-        raise Error::Unimplemented, "Unsupported bundle content: #{bundle.content}"
-      end
-
-      tlog_entries = bundle.verification_material.tlog_entries
-      raise Error::InvalidBundle, "Expected one tlog entry" if tlog_entries.size != 1
-
-      tlog_entry = tlog_entries.first
-
-      if media_type == BundleType::BUNDLE_0_1
-        unless tlog_entry.inclusion_promise
+    def validate_version!
+      case bundle_type
+      when BundleType::BUNDLE_0_1
+        unless verification_material.tlog_entries.all?(&:inclusion_promise)
           raise Error::InvalidBundle,
                 "bundle v0.1 requires an inclusion promise"
         end
-        if tlog_entry.inclusion_proof && !tlog_entry.inclusion_proof.checkpoint.envelope
+        if verification_material.tlog_entries.any? { |t| t.inclusion_proof&.checkpoint.nil? }
           raise Error::InvalidBundle,
                 "0.1 bundle contains an inclusion proof without checkpoint"
         end
       else
-        unless tlog_entry.inclusion_proof
+        unless verification_material.tlog_entries.all?(&:inclusion_proof)
           raise Error::InvalidBundle,
                 "must contain an inclusion proof"
         end
-        unless tlog_entry.inclusion_proof.checkpoint.envelope
+        unless verification_material.tlog_entries.all? { |t| t.inclusion_proof.checkpoint.envelope }
           raise Error::InvalidBundle,
                 "inclusion proof must contain a checkpoint"
         end
       end
 
-      entry = Sigstore::Transparency::LogEntry.from_proto(tlog_entry)
+      raise Error::InvalidBundle, "Expected one tlog entry" if verification_material.tlog_entries.size > 1
```

Comment:
is this legit?

Reply:
This is underspecified on the client side: sigstore-python currently chooses to limit its bundled to 1 tlog entry for simplicity reasons, but there's general requirement that clients do.

(The reason sigstore-python does it is because it's not clear whether multiple entries should be treated as a threshold, or "all or nothing.")

Reply:
sounds good -- we should get this specified in the client spec then, let's make an issue over there?

Reply:
https://github.com/sigstore/architecture-docs/issues/20